pkg/logstructured/logstructured.go

```diff
@@ -262,8 +268,85 @@ func (l *LogStructured) Update(ctx context.Context, key string, value []byte, re
 	return rev, updateEvent.KV, true, err
 }
 
+func (l *LogStructured) ttl(ctx context.Context) {
+	// vary naive TTL support
+	queue := workqueue.NewDelayingQueue()
+	ttlEventMap := &sync.Map{}
+	go func() {
+		for l.handleTTLEvents(ctx, queue, ttlEventMap) {
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			queue.ShutDown()
+			return
+		default:
+		}
+
+		for event := range l.ttlEvents(ctx) {
+			if event.Delete {
+				continue
+			}
+
+			expiredAt := time.Now().Add(time.Duration(event.KV.Lease) * time.Second)
+			eventKV := loadTTLEventKV(ttlEventMap, event.KV.Key)
+			if eventKV == nil {
+				storeTTLEventKV(ttlEventMap, event.KV.Key, event.KV.ModRevision, expiredAt)
+				queue.AddAfter(event.KV.Key, time.Duration(event.KV.Lease)*time.Second)
+			} else {
+				if event.KV.ModRevision > eventKV.modRevision {
+					storeTTLEventKV(ttlEventMap, event.KV.Key, event.KV.ModRevision, expiredAt)
+				}
+			}
+		}
+	}
+}
+
+func (l *LogStructured) handleTTLEvents(ctx context.Context, queue workqueue.DelayingInterface, ttlEventMap *sync.Map) bool {
+	key, shutdown := queue.Get()
+	if shutdown {
+		logrus.Info("TTL events work queue has shut down")
+		return false
+	}
+	defer queue.Done(key)
+
+	eventKV := loadTTLEventKV(ttlEventMap, key)
+	if eventKV == nil {
+		logrus.Errorf("Failed to find ttl event for key %v", key)
+		return true
+	}
+
+	if eventKV.expiredAt.After(time.Now()) {
+		logrus.Tracef("TTL event key %v has not expired yet, the latest expiration time is %v, requeuing", key, eventKV.expiredAt)
+		queue.AddAfter(key, eventKV.expiredAt.Sub(time.Now()))
```

Comment:
nit: use the idiomatic `Until`
```suggestion
		queue.AddAfter(key, time.Until(eventKV.expiredAt))
```